File: evaluate-examples.pl
#!/usr/bin/env perl

use v5.14;
use strict;
use warnings;
use utf8::all;
use autodie;

# Main

parse();

# Subroutines

sub parse {
    my @blocks;
    my $current_block = "";
    my $state = "out";

    while (my $line = <>) {
        if ($state eq "out") {
            if ($line eq "~~~ {.clojure}\n") {
                $state = "in";
                push @blocks, $current_block . $line;
                $current_block = "";
            }
            else {
                $current_block .= $line;
            }
        }
        elsif ($state eq "in") {
            if ($line eq "~~~\n") {
                $state = "out";
                push @blocks, evaluate($current_block);
                $current_block = $line;
            }
            elsif ($line eq "~~~ {.clojure}\n") {
                die "syntax error;"
            }
            else {
                $current_block .= $line;
            }
        }
    }

    push @blocks, $current_block;

    say join "", @blocks;
}

sub evaluate {
    my ($block) = @_;

    return "EVALUATED LOL\n";
}

File: evaluate-examples.clj
(defn evaluate [input]
  "\nLOL EVALUATED\n")

(defn parse [content]
  (let [lines (clojure.string/split content #"\n")
        lines# (count lines)]
    (clojure.string/join
      ""
      (loop [blocks []
             current-block ""
             state :out
             line# 0]
        (if (= line# lines#)
          (conj blocks (str current-block "\n"))
          (let [current-line (get lines line#)
                next-line# (inc line#)
                result-block (str current-block "\n" current-line)]
            (case state
              :out (if (= current-line "~~~ {.clojure}")
                     (recur (conj blocks result-block)
                            ""
                            :in
                            next-line#)
                     (recur blocks
                            result-block
                            state
                            next-line#))
              :in (if (= current-line "~~~")
                    (recur (conj blocks (evaluate result-block))
                           current-line
                           :out
                           next-line#)
                    (recur blocks
                           result-block
                           state
                           next-line#)))))))))

(let [file-name (first *command-line-args*)]
  (println "Parsing " file-name)
  (println (parse (slurp file-name))))
